File: tests/go/ice-test/result/point_result.go
package result

import (
	"context"

	ice "github.com/zjn-zjn/ice/sdks/go"
	icecontext "github.com/zjn-zjn/ice/sdks/go/context"
)

func init() {
	ice.RegisterLeaf("com.ice.test.result.PointResult",
		&ice.LeafMeta{
			Name:  "发放积分节点",
			Desc:  "用于发放积分奖励",
			Alias: []string{"point_result"},
		},
		func() any { return &PointResult{} })

	ice.RegisterLeaf("com.ice.test.result.PointResult2",
		&ice.LeafMeta{
			Name:  "发放积分节点2",
			Desc:  "另一个积分发放",
			Alias: []string{"point_result_2"},
		},
		func() any { return &PointResult2{} })

	ice.RegisterLeaf("com.ice.test.result.InitConfigResult",
		&ice.LeafMeta{
			Name:  "初始化配置节点",
			Desc:  "初始化roam中的配置",
			Alias: []string{"init_config_result"},
		},
		func() any { return &InitConfigResult{} })
}

// PointResult grants points to a user.
type PointResult struct {
	Key   string  `json:"key" ice:"name:用户ID键,desc:从roam中获取用户ID的键名"`
	Value float64 `json:"value" ice:"name:积分值,desc:要发放的积分数量"`
}

// DoResult implements the LeafResult interface.
func (p *PointResult) DoResult(ctx context.Context, roam *icecontext.Roam) bool {
	uid := roam.ValueDeep(p.Key).Int()
	if uid == 0 || p.Value <= 0 {
		return false
	}

	res := sendService.SendPoint(uid, p.Value)
	roam.Put("SEND_POINT", res)
	return res
}

// PointResult2 is another variant of point granting.
type PointResult2 struct {
	Key   string  `json:"key" ice:"name:用户ID键,desc:从roam中获取用户ID的键名"`
	Value float64 `json:"value" ice:"name:积分值,desc:要发放的积分数量"`
}

// DoResult implements the LeafResult interface.
func (p *PointResult2) DoResult(ctx context.Context, roam *icecontext.Roam) bool {
	uid := roam.ValueDeep(p.Key).Int()
	if uid == 0 || p.Value <= 0 {
		return false
	}

	res := sendService.SendPoint(uid, p.Value)
	roam.Put("SEND_POINT", res)
	return res
}

// InitConfigResult initializes config in roam.
type InitConfigResult struct {
	ConfigKey   string `json:"configKey" ice:"name:配置键,desc:要设置的配置键名"`
	ConfigValue string `json:"configValue" ice:"name:配置值,desc:要设置的配置值"`
}

// DoResult implements the LeafResult interface.
func (i *InitConfigResult) DoResult(ctx context.Context, roam *icecontext.Roam) bool {
	if i.ConfigKey == "" {
		return false
	}
	roam.Put(i.ConfigKey, i.ConfigValue)
	return true
}

File: tests/python/ice-test/result/point_result.py
"""Point result leaf nodes for testing."""

from typing import Annotated

import ice
from ice import IceField, Roam
from service.send_service import send_service


@ice.leaf(
    "com.ice.test.result.PointResult",
    name="发放积分节点",
    desc="用于发放积分奖励",
    alias=["point_result"],
)
class PointResult:
    """Grants points to a user."""
    key: Annotated[str, IceField(name="用户ID键", desc="从roam中获取用户ID的键名")] = ""
    value: Annotated[float, IceField(name="积分值", desc="要发放的积分数量")] = 0.0
    
    def do_roam_result(self, roam: Roam) -> bool:
        uid_val = roam.get_multi(self.key)
        if uid_val is None or self.value <= 0:
            return False
        
        try:
            uid = int(uid_val)
        except (ValueError, TypeError):
            return False
        
        res = send_service.send_point(uid, self.value)
        roam.put("SEND_POINT", res)
        return res


@ice.leaf(
    "com.ice.test.result.PointResult2",
    name="发放积分节点2",
    desc="另一个积分发放",
    alias=["point_result_2"],
)
class PointResult2:
    """Another variant of point granting."""
    key: Annotated[str, IceField(name="用户ID键", desc="从roam中获取用户ID的键名")] = ""
    value: Annotated[float, IceField(name="积分值", desc="要发放的积分数量")] = 0.0
    
    def do_roam_result(self, roam: Roam) -> bool:
        uid_val = roam.get_multi(self.key)
        if uid_val is None or self.value <= 0:
            return False
        
        try:
            uid = int(uid_val)
        except (ValueError, TypeError):
            return False
        
        res = send_service.send_point(uid, self.value)
        roam.put("SEND_POINT", res)
        return res


@ice.leaf(
    "com.ice.test.result.InitConfigResult",
    name="初始化配置节点",
    desc="初始化roam中的配置",
    alias=["init_config_result"],
)
class InitConfigResult:
    """Initializes config in roam."""
    configKey: Annotated[str, IceField(name="配置键", desc="要设置的配置键名")] = ""
    configValue: Annotated[str, IceField(name="配置值", desc="要设置的配置值")] = ""
    
    def do_roam_result(self, roam: Roam) -> bool:
        if not self.configKey:
            return False
        roam.put(self.configKey, self.configValue)
        return True
